presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSource.java

```diff
@@ -0,0 +1,389 @@
+        this.columnHandles = ImmutableList.copyOf(columnHandles);
+        this.session = session;
+        this.objectMapper = objectMapper;
+    }
+
+    private static Double parseDouble(String value)
+    {
+        try {
+            return Double.valueOf(value);
+        }
+        catch (NumberFormatException ne) {
+            switch (value) {
+                case PINOT_INFINITY:
+                case PINOT_POSITIVE_INFINITY:
+                    return PRESTO_INFINITY;
+                case PINOT_NEGATIVE_INFINITY:
+                    return PRESTO_NEGATIVE_INFINITY;
+            }
+            throw new PinotException(PinotErrorCode.PINOT_DECODE_ERROR, Optional.empty(), "Cannot decode double value from pinot " + value, ne);
+        }
+    }
+
+    private void setValue(Type type, BlockBuilder blockBuilder, String value)
+    {
+        if (value == null) {
+            blockBuilder.appendNull();
+            return;
+        }
+        boolean handled = true;
+        if (type instanceof FixedWidthType) {
+            completedBytes += ((FixedWidthType) type).getFixedSize();
+            if (type instanceof BigintType) {
+                type.writeLong(blockBuilder, parseDouble(value).longValue());
+            }
+            else if (type instanceof IntegerType) {
+                blockBuilder.writeInt(parseDouble(value).intValue());
+            }
+            else if (type instanceof TinyintType) {
+                blockBuilder.writeByte(parseDouble(value).byteValue());
+            }
+            else if (type instanceof SmallintType) {
+                blockBuilder.writeShort(parseDouble(value).shortValue());
+            }
+            else if (type instanceof BooleanType) {
+                type.writeBoolean(blockBuilder, Boolean.valueOf(value));
+            }
+            else if (type instanceof DecimalType || type instanceof DoubleType) {
+                type.writeDouble(blockBuilder, parseDouble(value).doubleValue());
+            }
+            else if (type instanceof TimestampType) {
+                type.writeLong(blockBuilder, Long.valueOf(value));
+            }
+            else {
+                handled = false;
+            }
+        }
+        else if (type instanceof VarcharType) {
+            Slice slice = Slices.utf8Slice(value);
+            blockBuilder.writeBytes(slice, 0, slice.length()).closeEntry();
+            completedBytes += slice.length();
+        }
+        else {
+            handled = false;
+        }
+        if (!handled) {
+            throw new PinotException(PinotErrorCode.PINOT_UNSUPPORTED_COLUMN_TYPE, Optional.empty(), "type '" + type + "' not supported");
+        }
+    }
+
+    private void setValuesForGroupby(List<BlockBuilder> blockBuilders, List<Type> types, int numGroupByClause, JsonNode group, String[] values)
+    {
+        for (int k = 0; k < group.size(); k++) {
+            setValue(types.get(k), blockBuilders.get(k), group.get(k).asText());
```

Comment:
Use `stream()` and `foreach()`. In general, we prefer lambdas to classic for loops for these kinds of cases.

Reply:
Its more complicated to use such when iterating over 3 parallel lists here. Possible but more verbose with IntStream/IntRange and like. Programmers can grok for loops pretty well.

Reply:
I was suggesting the use of `Streams.zip` since we are using Guava. However, I didn't notice the third list which makes using that cumbersome. Keep it as is then